php-checks/src/main/java/org/sonar/php/checks/FileWithSymbolsAndSideEffectsCheck.java

```diff
@@ -19,117 +19,99 @@
-
-  private static class File {
-    boolean hasSymbol = false;
-    boolean hasSideEffects = false;
-    boolean inDeclaration = false;
-    boolean hasIssue = false;
-
-    public void reset() {
-      hasSymbol = false;
-      hasSideEffects = false;
-      hasIssue = false;
-      inDeclaration = false;
-    }
-  }
+public class FileWithSymbolsAndSideEffectsCheck extends PHPVisitorCheck {
 
-  private static AstNodeType[] DECLARATIONS = {
-    PHPGrammar.CLASS_DECLARATION,
-    PHPGrammar.FUNCTION_DECLARATION,
-    PHPGrammar.INTERFACE_DECLARATION
-  };
+  public static final String KEY = "S2036";
+  private static final String MESSAGE = "Refactor this file to either declare symbols or cause side effects, but not both.";
 
-  private static AstNodeType[] SIDE_EFFECTS_STATEMENT = {
-    PHPGrammar.YIELD_STATEMENT,
-    PHPGrammar.ECHO_STATEMENT,
-    PHPGrammar.INLINE_HTML,
-    PHPGrammar.UNSET_VARIABLE_STATEMENT,
-    PHPGrammar.EXPRESSION_STATEMENT
-  };
+  private boolean fileHasSymbol;
+  private boolean fileHasSideEffect;
 
   @Override
-  public void init() {
-    subscribeTo(DECLARATIONS);
-    subscribeTo(SIDE_EFFECTS_STATEMENT);
-  }
+  public void visitCompilationUnit(CompilationUnitTree tree) {
+    fileHasSymbol = false;
+    fileHasSideEffect = false;
 
-  private File currentFile = new File();
+    super.visitCompilationUnit(tree);
 
-  @Override
-  public void visitFile(@Nullable AstNode astNode) {
-    currentFile.reset();
+    if (fileHasSymbol && fileHasSideEffect) {
+      context().newIssue(KEY, MESSAGE);
+    }
   }
 
   @Override
-  public void visitNode(AstNode astNode) {
-    if (astNode.is(DECLARATIONS)) {
-      currentFile.inDeclaration = true;
+  public void visitClassDeclaration(ClassDeclarationTree tree) {
+    if (tree.is(Kind.CLASS_DECLARATION, Kind.INTERFACE_DECLARATION)) {
+      fileHasSymbol = true;
+      // do not enter inside class declaration
+    }
+  }
 
-      if (!currentFile.hasSymbol) {
-        currentFile.hasSymbol = true;
-      }
+  @Override
+  public void visitFunctionDeclaration(FunctionDeclarationTree tree) {
+    fileHasSymbol = true;
+    // do not enter inside function declaration
+  }
 
-      // Do not report "side-effects" statement within a declaration
-    } else if (!currentFile.inDeclaration && !currentFile.hasSideEffects && astNode.is(SIDE_EFFECTS_STATEMENT) && !isExcluded(astNode)) {
-      currentFile.hasSideEffects = true;
-    }
+  @Override
+  public void visitYieldStatement(YieldStatementTree tree) {
+    super.visitYieldStatement(tree);
+    fileHasSideEffect = true;
+  }
 
-    if (!currentFile.hasIssue && currentFile.hasSymbol && currentFile.hasSideEffects) {
-      getContext().createFileViolation(this, "Refactor this file to either declare symbols or cause side effects, but not both.");
-      currentFile.hasIssue = true;
+  @Override
+  public void visitInlineHTML(InlineHTMLTree tree) {
+    boolean isClosingTag = "?>".equals(tree.inlineHTMLToken().text().trim());
```

Comment:
@vilchik-elena I'm not sure where this comes from. If there are spaces after a closing tag, then this will trigger some output and should probably be considered as a side effect.